File: token-client-core/src/main/java/no/nav/security/token/support/client/core/oauth2/TokenExchangeClient.java
package no.nav.security.token.support.client.core.oauth2;

import no.nav.security.token.support.client.core.http.OAuth2HttpClient;

import java.util.LinkedHashMap;
import java.util.Map;

import static no.nav.security.token.support.client.core.OAuth2ParameterNames.*;

public class TokenExchangeClient extends AbstractOAuth2TokenClient<TokenExchangeGrantRequest> {

    public TokenExchangeClient(OAuth2HttpClient oAuth2HttpClient) {
        super(oAuth2HttpClient);
    }

    @Override
    protected Map<String, String> formParameters(TokenExchangeGrantRequest grantRequest) {
        Map<String, String> formParameters = new LinkedHashMap<>();
        var tokenExchangeProperties = grantRequest.getClientProperties().getTokenExchange();
        formParameters.put(SUBJECT_TOKEN_TYPE, tokenExchangeProperties.subjectTokenType());
        formParameters.put(SUBJECT_TOKEN, grantRequest.getSubjectToken());
        formParameters.put(AUDIENCE, tokenExchangeProperties.getAudience());
        if (tokenExchangeProperties.getResource() != null && !tokenExchangeProperties.getResource().isEmpty()) {
            formParameters.put(RESOURCE, tokenExchangeProperties.getResource());
        }
        return formParameters;
    }
}

File: token-client-core/src/main/kotlin/no/nav/security/token/support/client/core/oauth2/TokenExchangeClient.kt
package no.nav.security.token.support.client.core.oauth2

import no.nav.security.token.support.client.core.ClientProperties.TokenExchangeProperties.Companion.SUBJECT_TOKEN_TYPE_VALUE
import no.nav.security.token.support.client.core.OAuth2ParameterNames.AUDIENCE
import no.nav.security.token.support.client.core.OAuth2ParameterNames.RESOURCE
import no.nav.security.token.support.client.core.OAuth2ParameterNames.SUBJECT_TOKEN
import no.nav.security.token.support.client.core.OAuth2ParameterNames.SUBJECT_TOKEN_TYPE
import no.nav.security.token.support.client.core.http.OAuth2HttpClient

class TokenExchangeClient(oAuth2HttpClient : OAuth2HttpClient) : AbstractOAuth2TokenClient<TokenExchangeGrantRequest>(oAuth2HttpClient) {

    override fun formParameters(grantRequest : TokenExchangeGrantRequest) =
        with(grantRequest)  {
            HashMap<String, String>().apply {
                clientProperties.tokenExchange?.run {
                    put(SUBJECT_TOKEN_TYPE, SUBJECT_TOKEN_TYPE_VALUE)
                    put(SUBJECT_TOKEN,subjectToken)
                    put(AUDIENCE, audience)
                    resource?.takeIf { it.isNotEmpty() }?.let { put(RESOURCE, it) }
                }
            }
        }
}